File: 3449.maximize-the-minimum-game-score.java
#
# @lc app=leetcode id=3449 lang=java
#
# [3449] Maximize the Minimum Game Score
#

# @lc code=start
class Solution {
    public long maxScore(int[] points, int m) {
        int n = points.length;
        long low = 0;
        long high = 0;
        for (int p : points) {
            high = Math.max(high, 1L * p * m);
        }
        while (low < high) {
            long mid = low + (high - low + 1) / 2;
            if (canAchieve(points, (long) m, mid)) {
                low = mid;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
    private boolean canAchieve(int[] points, long m, long x) {
        int n = points.length;
        long[] need = new long[n];
        long[] suffixMax = new long[n];
        suffixMax[n - 1] = (x + points[n - 1] - 1L) / points[n - 1];
        need[n - 1] = suffixMax[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            need[i] = (x + points[i] - 1L) / points[i];
            suffixMax[i] = Math.max(need[i], suffixMax[i + 1]);
        }
        long total = 0;
        long depLeft = 0;
        for (int i = 0; i < n; i++) {
            long depRight = (i < n - 1 ? suffixMax[i + 1] : 0L);
            long k = Math.max(need[i], depLeft + depRight);
            total += k;
            if (total > m) return false;
            depLeft = Math.max(0L, k - depRight);
        }
        return true;
    }
}
# @lc code=end

File: 3449.maximize-the-minimum-game-score.go
#
# @lc app=leetcode id=3449 lang=golang
#
# [3449] Maximize the Minimum Game Score
#

# @lc code=start
func maxScore(points []int, m int) int64 {
    n := len(points)
    minP := int64(points[0])
    for _, p := range points {
        pp := int64(p)
        if pp < minP {
            minP = pp
        }
    }
    left, right := int64(0), minP * int64(m)
    for left < right {
        mid := left + (right - left + 1) / 2
        if canAchieve(points, mid, int64(m), n) {
            left = mid
        } else {
            right = mid - 1
        }
    }
    return left
}

func canAchieve(points []int, mid, M int64, n int) bool {
    f := make([]int64, n)
    f[0] = (mid + int64(points[0]) - 1) / int64(points[0])
    for i := 1; i < n; i++ {
        need := (mid + int64(points[i]) - 1) / int64(points[i])
        f[i] = max(need, f[i-1] - 1)
    }
    total := int64(0)
    for i := 0; i < n; i++ {
        total += f[i]
        if total > M {
            return false
        }
    }
    // Backward pass with incremental update to catch extra costs
    currb := f[n-1]
    for i := n-2; i >= 0; i-- {
        need := (mid + int64(points[i]) - 1) / int64(points[i])
        currb = max(need, currb - 1)
        if currb > f[i] {
            total += currb - f[i]
            if total > M {
                return false
            }
            f[i] = currb
        }
    }
    return total <= M
}

func max(a, b int64) int64 {
    if a > b {
        return a
    }
    return b
}
# @lc code=end